Fix #12908 Add ICM-42688-P IMU to list of IMUs with overflow protection

tbolin opened the pull request:
What it says in the description.
The ICM-42688-P have been verified to have overflow protection in https://github.com/betaflight/betaflight/issues/12908.
Still have to find a working 42605 to test with.
The ICM-42605 is preliminary added to the list of gyros with overflow protection, since it's hard to test if a gyros has overflow protection if Betaflight's overflow protection is active.

**testing instructions**

* Have a FC capable of running betaflight 4.5, has a ICM-42605 IMU, and can log
* Flash this PR 
* Go to cli and set `blackbox_mode = ALWAYS`, `debug_mode = GYRO_RAW` and save
* Wait for FC to reboot and the gyro to calibrate, then flick the FC quickly a couple of times.
* Go to cli and set `blackbox_mode = NORMAL` and save
* Extract the log and look where the gyro traces reach 2000 dps. If the gyro has overflow protection the rate should remain at around 
2000 dps for a short while. If the gyros is subject to overflow, the rate should suddenly change sign.

nerdCopter commented:
very few with 42605 defined:
```
DARWINF4SX1280HD
GEPRCF411SX1280
MAMBAH743
MAMBAH743_2022B
MATEKH743
NEUTRONRCF411AIO
NEUTRONRCF411SX1280
SPRACINGH7RF
TAKERF722SE
ZEEZF7V3
```
maybe @hydra has one of his on hand for testing, or maybe already knows the answer.

tbolin commented:
The annoying thing is that I have a MATEK H743-SLIM V3 board with both the 42688-P and the 42605 on it, but for some reason the 42605 isn't detected.

haslinghuis commented:
My MatekH743 only shows MPU6000

tbolin commented:
I tried changing the config so the ICM-42688-P was the second gyro and the 42605 the first. It detected the 42688 as the second gyro, but not the 42605.

Quick-Flash commented:
> My MatekH743 (SLIM V1.5) only shows MPU6000
> 
> ![image](https://user-images.githubusercontent.com/8344830/258649510-bc20f6e4-797e-467a-9ff4-94a2d1aa9926.png)
> 
> After `set gyro_to_use = SECOND` it just reverts to `FIRST` after save. So guess we have to check if that happens with other boards with dual gyro's too.

I have the same problem, I used to be able to get the second gyro to work on my matek board, but now regardless of what I do I cannot get it to work even when flashing an older version of BF.

hydra commented:
> maybe @hydra has one of his on hand for testing, or maybe already knows the answer.

I believe Invesense fixed it on the ICM42605, which is very similar to the ICM42688P. I do have FC's here with 42605's on.  Likely it'll be quicker to ask Invesense/TDK directly than wait for me to find time to verify things as I'm not sure if I've got one in a flyable rig right now.

If the procedure for verification can be documented that would help others get the answer too.

tbolin commented:
> I believe Invesense fixed it on the ICM42605, which is very similar to the ICM42688P. I do have FC's here with 42605's on. Likely it'll be quicker to ask Invesense/TDK directly than wait for me to find time to verify things as I'm not sure if I've got one in a flyable rig right now.

I think it's easier to check if the IMU suffers from overflow if the FC isn't connected to anything, but contacting Ivensense might be a good idea anyway.

> If the procedure for verification can be documented that would help others get the answer too.

Good point. I have added testing instruction to the first post. That should also be a good way to check that I haven't misunderstood how the gyro overflow issue works.

nerdCopter commented:
logs of this PR attached. shaken wildly by hand only.
`23ae8fa41` FOXERRF722V4 ICM42688P
[BTFL_BLACKBOX_LOG_20230816_132055_FOXEERF722V4.BBL.zip](https://github.com/betaflight/betaflight/files/12362430/BTFL_BLACKBOX_LOG_20230816_132055_FOXEERF722V4.BBL.zip)

![image](https://github.com/betaflight/betaflight/assets/56646290/995c3d69-85f8-4000-b0b0-e3bb6bec2100)

haslinghuis commented:
Very small difference between gyro and unfiltered gyro.

tbolin commented:
@haslinghuis 
> Very small difference between gyro and unfiltered gyro.
Debug mode GYRO_SCALED seems to be no longer needed?

Unfiltered gyro and *GYRO_SCALED* is intended to be identical. I haven't double checked if they actually are, so *GYRO_SCALED* should be redundant.
*DEBUG_DSHOT_RPM_TELEMETRY* should also be the same as the RPM data field, at least after
https://github.com/betaflight/betaflight/pull/13012 gets merged.

tbolin commented:
the PR is still under way. I just have to find a board with a ICM42605 to test.

hydra commented:
> the PR is still under way. I just have to find a board with a ICM42605 to test.

I have some H7RF's and other H7 based FC's that I fitted ICM42605's to (they're pin compatible with ICM42688P) so if you'd like me to test some code let me know and please give me instructions on how to test it.

tbolin commented:
@hydra 
> I have some H7RF's and other H7 based FC's that I fitted ICM42605's to (they're pin compatible with ICM42688P) so if you'd like me to test some code let me know and please give me instructions on how to test it.

I would be very grateful if you could test the PR. The instructions for testing are in the first post. Let me know if there are any details missing.

ctzsnooze commented:
@tbolin I wonder if we can take out the ICM42605 since we have no data on it, and then merge this PR into 4.5, knowing that the ICM-42688-P does have protection?   We could then perhaps make a separate PR for the ICM42605 once we know it is OK.

nerdCopter reviewed (approved):
* approving with assumption that ICM42605 would also have overflow protection.  would be easy to ask some TDK forum possibly. or if desired, as ctsnooze said, could leave it out get this merged. both gyro marketing as "SmartMotion" and both no mention of overflow in datasheets.

ctzsnooze commented:
We can merge this only once the ICM42605 situation is absolutely clear.  If that cannot be ascertained with a high level of confidence, then please remove the reference to the ICM42605 so that we can merge this.

tbolin commented:
> We can merge this only once the ICM42605 situation is absolutely clear. If that cannot be ascertained with a high level of confidence, then please remove the reference to the ICM42605 so that we can merge this.

Agreed. I removed the ICM42605 from the list in this PR and renamed the PR itself to reflect the changes so that it can finally be merged. If someone finds a FC with a working 42605 it's trivial to add that in a later PR.

ctzsnooze commented:
Thank you!  Will get merged asap.

hydra commented:
> > We can merge this only once the ICM42605 situation is absolutely clear. If that cannot be ascertained with a high level of confidence, then please remove the reference to the ICM42605 so that we can merge this.
> 
> Agreed. I removed the ICM42605 from the list in this PR and renamed the PR itself to reflect the changes so that it can finally be merged. If someone finds a FC with a working 42605 it's trivial to add that in a later PR.

any chance you could prep the PR for the 42605 and link it from here?  Its on my list to test the 42605 but ive been taking a break.  Others with the 42605 would also then have test binaries and a place to report success/failure and we wouldn't loose visibility or tracking of the current situation.

tbolin commented:
> any chance you could prep the PR for the 42605 and link it from here? Its on my list to test the 42605 but ive been taking a break. Others with the 42605 would also then have test binaries and a place to report success/failure and we wouldn't loose visibility or tracking of the current situation.

Good idea. Will do.

tbolin commented:
@hydra  #13194 is now up and have the ICM-42605 added to the list of gyros with overflow protection
